map_test.go

```diff
@@ -1726,69 +1726,68 @@ func TestMapPinningIncompatible(t *testing.T) {
 	tmp := testutils.TempBPFFS(t)
 	c := qt.New(t)
 
-	spec := &MapSpec{
+	m1, err := NewMapWithOptions(&MapSpec{
 		Name:       "test",
 		Type:       Hash,
 		KeySize:    4,
 		ValueSize:  4,
 		MaxEntries: 1,
 		Pinning:    PinByName,
-	}
-
-	m1, err := NewMapWithOptions(spec, MapOptions{PinPath: tmp})
+	}, MapOptions{PinPath: tmp})
 	if err != nil {
 		t.Fatal("Can't create map:", err)
 	}
 	defer m1.Close()
-	pinned := m1.IsPinned()
-	c.Assert(pinned, qt.IsTrue)
 
-	_, err = m1.Info()
-	c.Assert(err, qt.IsNil)
-
-	if err := m1.Put(uint32(0), uint32(42)); err != nil {
-		t.Fatal("Can't write value:", err)
-	}
-	spec.KeySize = 8
-	tc := map[string]*MapSpec{
-		"KeySize":    {Name: spec.Name, Type: spec.Type, KeySize: 8, ValueSize: spec.ValueSize, MaxEntries: spec.MaxEntries, Pinning: spec.Pinning},
-		"ValueSize":  {Name: spec.Name, Type: spec.Type, KeySize: spec.ValueSize, ValueSize: 8, MaxEntries: spec.MaxEntries, Pinning: spec.Pinning},
-		"Type":       {Name: spec.Name, Type: Array, KeySize: spec.ValueSize, ValueSize: spec.ValueSize, MaxEntries: spec.MaxEntries, Pinning: spec.Pinning},
-		"MaxEntries": {Name: spec.Name, Type: spec.Type, KeySize: spec.ValueSize, ValueSize: spec.ValueSize, MaxEntries: 2, Pinning: spec.Pinning},
-		"Flags":      {Name: spec.Name, Type: spec.Type, KeySize: spec.ValueSize, ValueSize: spec.ValueSize, MaxEntries: spec.MaxEntries, Pinning: spec.Pinning, Flags: unix.BPF_F_RDONLY_PROG},
+	tests := []struct {
+		name string
+		spec *MapSpec
+		test func(*qt.C, *IncompatibleMapError)
+	}{
+		{"Type", &MapSpec{Name: "test", Type: Array, KeySize: 4, ValueSize: 4, MaxEntries: 1, Pinning: PinByName},
```

Comment:
This refactoring makes it harder to figure out which field has changed due to a lot of repetition. The previous code also has repetition but it's simpler to parse.

I'd probably go for a field `func(*MapSpec)` and then set `func(ms *MapSpec) { ms.Type = Array }`. Then it's clear what we are changing.

Reply:
Good idea! I feel like we have many more tests that could take this approach. :)